source/common/router/header_parser.cc

```diff
@@ -11,18 +17,70 @@ HeaderFormatterPtr parseInternal(const envoy::api::v2::HeaderValueOption& header
   const std::string& format = header_value_option.header().value();
   const bool append = PROTOBUF_GET_WRAPPED_OR_DEFAULT(header_value_option, append, true);
 
-  if (format.find("%") == 0) {
-    const size_t last_occ_pos = format.rfind("%");
-    if (last_occ_pos == std::string::npos || last_occ_pos <= 1) {
-      throw EnvoyException(fmt::format("Incorrect header configuration. Expected variable format "
-                                       "%<variable_name>%, actual format {}",
-                                       format));
+  std::vector<HeaderFormatterPtr> formatters;
+
+  const absl::string_view format_view(format);
+  std::string buf;
+  buf.reserve(format_view.size());
+
+  size_t prev = 0;
```

Comment:
Can we get a bunch more comments on this code. I personally find it very hard to read this type of logic quickly and comments can help a lot.